### DISCUSSION TITLE
Why is storySource necessary?

### DISCUSSION THREAD

**Discussion by joriswitteman** [daimresearch/sb-addon-permutation-table#22] [General]
I feel very strange having to fill in `source` and `importPath`. Aren't those already defined in the Story?

Couldn't this add-on derive its permutations from the Story's `context`, which contains all the `argTypes` ?

Couldn't a wrapper around the story, repeat the same story with all the different Controls values, in a table like it does now?

For example, the following code implements a Decorator that works for me in displaying all permutations in a single Story, without any exotic custom parameters:

```tsx
  import { StoryFn, StoryContext } from '@storybook/react'

  export const AllVariationsDecorator = (Story: StoryFn, storyContext: StoryContext) => {
    const { argTypes, args: defaultArgs } = storyContext
  
    if (argTypes) {
      // Extract possible arguments and their options from Story context

      const allArgs: Record<string, unknown[]> = {}
      const allVariations: Record<string, unknown>[] = []
  
      for (const key in argTypes) {
        const argType = argTypes[key]
        const valueType = String(argType.type?.name)
  
        switch (valueType) {
          case 'boolean':
            allArgs[key] = [true, false]
          case 'string':
            if (argType.options) {
              // Limit dropdowns to top 4 options for performance
              allArgs[key] = argType.options.slice(0, 4)
            }
        }
      }
  
      // Compose an array of all possible combinations of allArgs,
      // as argument key/value objects

      allVariations.push({})
  
      for (const argKey in allArgs) {
        for (const oppositeArgKey in allArgs) {
          if (argKey === oppositeArgKey) continue

          for (const argOption of allArgs[argKey]) {
            for (const oppositeArgOption of allArgs[oppositeArgKey]) {
              const argsVariation = {
                [argKey]: argOption,
                [oppositeArgKey]: oppositeArgOption,
              }
  
              allVariations.push(argsVariation)
            }
          }
        }
      }
  
      // Render every variation of the Story based on possible arg values
      return (
        <>
          {allVariations.map(variationArgs => (
            <Story args={{...defaultArgs, ...variationArgs}} />
          ))}
        </>
      )
    } else {
      return <Story />
    }
  }
```

**Comment by dongyunlee-daim** [comment]
@joriswitteman This seems like a really great idea. Can you tell me what your working environment is like, and what version of Storybook do you use?

**Reply by joriswitteman** [reply]
Sure. We're using Storybook 7.0.24 currently, with React / Next.js.

Our environment is a large design system of 50+ components, some with up to 12 controls, of which some have up to 8 options (as a dropdown). This unfortunately makes for thousands of permutations, hence my suggestion on Discord to work on some kind of progressive disclosure (perhaps reveal on scroll?).

Our use case for a permutations table is actually to add component variations to test coverage, when running the Storybook test runner and shooting snapshots. As I fear the test will hang while rendering thousands of permutations, I am now considering moving away from above decorator, to a run-time clicker that simply tries out all the controls and shoots snapshots one by one instead.

**Reply by dongyunlee-daim** [reply]
Thanks for sharing! our team checked this issues for serval days. In conclusion, it may be possible to remove the storySource completely, but for now, it would require a delicate approach to use components that require children (like a regular html button) without a storySource. It would also be a very big change for our project and would require a lot of effort and time, including rewriting the README. However, we're also committed to zero-config when using our addon, so we're going to invest heavily in this agenda.

Furthermore, while digging into this discussion, there is a glimmer of hope that our addon can be used outside of React projects (related: #20 ). We'll share any further developments here.

**Reply by dongyunlee-daim** [reply]
@joriswitteman  Thanks to your sincere interest and help, We have succeeded in removing the storySource from my setup. Not only that, but also made encouraging progress in removing dependencies on other libraries. Once again, thank you for your interest in our project!
